driver/ixsystems/iscsi.py

```diff
@@ -141,7 +146,49 @@ def remove_export(self, context, volume):
         """
         pass
 
+    def check_connection(self):
+        # connection safety check for #27
+        if ix_utils.parse_truenas_version(self.common._system_version())[1] in ('12.0', '13.0'):
+            LOG.debug("Tanable: %s" % str(self.common._tunable()))
+            tunable = self.common._tunable()
+            # default value from Truenas 12 kern.cam.ctl.max_ports 256, kern.cam.ctl.max_luns 1024
+            lunallowed = 256
+            for item in tunable:
+                if (item.get('enabled') and
+                        item.get('var') == 'kern.cam.ctl.max_luns' and
+                        str(item.get('value')).isnumeric()):
+                    lunallowed = min(lunallowed, int(item['value']))
+                if (item.get('enabled') and
+                        item.get('var') == 'kern.cam.ctl.max_ports' and
+                        str(item.get('value')).isnumeric()):
+                    lunallowed = min(lunallowed, int(item['value']))
+            LOG.debug("Tunable OS max_luns/max_ports: %s" % lunallowed)
+
+        # check cinder driver already loaded before executing upstream code
+        if (len(cinderapi.CONF.list_all_sections()) > 0):
+            ctx = context.get_admin_context()
+            ctx.__setattr__("read_deleted", "no")
+            ctx.__setattr__("project_only", "True")
+            vols = cinderapi.volume_get_all(ctx)
+            attached_truenas_vol_count = len([vol for vol in vols
+                                              if vol.host.find("@ixsystems-iscsi#") > 0 and vol.attach_status == 'attached'])
+            if (attached_truenas_vol_count >= lunallowed):
+                LOG.error("Maximum lun/port limitation reached. Change kern.cam.ctl.max_luns and "
+                          + "kern.cam.ctl.max_ports in tunable settings to allow more lun attachments.")
+                return False
+        return True
+
```

Comment:
Compare attached Truenas cinder volumes and and tunable parameter kern.cam.ctl.max_ports and kern.cam.ctl.max_luns, if attached Truenas cinder volumes count > tunable parameter kern.cam.ctl.max_ports and kern.cam.ctl.max_luns, raise exception.